pulsar-client/src/main/java/org/apache/pulsar/client/impl/TopicMessageImpl.java

```diff
@@ -32,11 +32,15 @@
 
     private final Message<T> msg;
     private final TopicMessageIdImpl messageId;
+    // consumer if this message is received by that consumer
+    ConsumerImpl receivedByconsumer;
 
     TopicMessageImpl(String topicPartitionName,
                      String topicName,
-                     Message<T> msg) {
+                     Message<T> msg,
+                     ConsumerImpl receivedByConsumer) {
         this.topicPartitionName = topicPartitionName;
+        this.receivedByconsumer = receivedByConsumer;
```

Comment:
can we add an assertion that `receivedByConsumer` is non null ?

Reply:
`receivedByConsumer` can be null if you want to create a message without consumer for some other purpose. so, it's intentional.